src/core/server/saved_objects/migrationsv2/state_action_machine.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/* eslint-disable prettier/prettier */
+
+export interface ControlState {
+  controlState: string;
+}
+
+/**
+ * A state-action machine for performing Saved Object Migrations.
+ *
+ * Based on https://www.microsoft.com/en-us/research/uploads/prod/2016/12/Computation-and-State-Machines.pdf
+ *
+ * The state-action machine defines it's behaviour in steps. Each step is a
+ * transition from a state s_i to the state s_i+1 caused by an action a_i.
+ *
+ * s_i   -> a_i -> s_i+1
+ * s_i+1 -> a_i+1 -> s_i+2
+ *
+ * Given a state s1, `next(s1)` returns the next action to execute. Actions are
+ * asynchronous, once the action resolves, we can use the action response to
+ * determine the next state to transition to as defined by the function
+ * `model(state, response)`.
+ *
+ * We can then loosely define a step as:
+ * s_i+1 = model(s_i, await next(s_i)())
+ *
+ * When there are no more actions returned by `next` the state-action machine
+ * terminates.
+ *
+ * @param initialState The initial state with which to start the state action
+ * machine
+ * @param next A function which given the current state, returns a thunk which
+ * is the next action to perform. If next returns null the state action machine
+ * terminates.
+ * @param model A function which given the current state and the response of
+ * the action thunk, returns a new state
+ * @param onStepComplete A callback functions which is called after every
+ * completed step
+ */
+export async function stateActionMachine<
+  S extends ControlState,
+>(
+  initialState: S,
+  // It would be nice to use generics to enforce that model should accept all
+  // the types of responses that actions could return. But seems to be
+  // impossible because of https://github.com/microsoft/TypeScript/issues/13995#issuecomment-477978591
+  next: (state: S) => (() => Promise<unknown>) | null,
+  model: (state: S, res: any) => S,
+) {
+  let state = initialState;
+  let controlStateStepCounter = 0;
+  let nextAction = next(state);
+
+  while (nextAction != null) {
+    // Perform the action that triggers the next step
+    const actionResult = await nextAction();
```

Comment:
Just want to validate my assumptions/knowledge here with a question...

From what I can tell, if `nextAction` throws / rejects, it doesn't appear to be caught anywhere. It seems to bubble up to `KibanaMigrator.runMigrations` which is called directly by `SavedObjectsService.start` and then finally caught in `Root.start` which will log the fatal error and shutdown Kibana.

I think that is ok, since we are only expecting to throw or reject errors that we aren't sure if we can automatically recover from. In case there is a class of errors that is recoverable that we aren't retrying in the code, the customer should still be able to successfully retry by restarting Kibana, correct?

Reply:
Yes. The way I've been thinking about it is that there are expected failures and exceptions. Expected failures are scenarios which we anticipate encountering under normal circumstances (like if we clone an index and receive index already exists, it's expected cause we anticipate multiple instances participating in the migration).

Everything else is an exception and the only way to recover from it is to retry the whole migration process from the beginning. I'm not 100% sure what the best option for retrying these exceptions. The alternative to crashing the process is to just restart the migrationStateMachine in an endless loop. Because kibana will almost always be restarted in a production environment it might not make such a big difference to users. The downside of restarting the process is there's additional overhead in restarting _everything_. The pro of restarting the process is that some orchestration frameworks will pick up a bootlooping process (e.g. Kubernetes CrashLoopBackOff). But I don't know if the same would happen if a process runs but always fails the liveness check within the timeout. I'll ask some internal teams for their input.

Reply:
We're not doing any kind of bootloop detection on Cloud. But their feedback was that most stack components will try for a limited amount of time (5 / 10 minutes) and then crash. This is so that failures that quickly resolve by themselves such as temporary performance or network degradation doesn't cause a reboot immediately and also somewhat rate-limits how quickly a stack component will bootloop.

Reply:
Works for me. Thanks for checking with them. 👍